add cross domain scoping

auhlig opened the pull request:
As described in https://github.com/improbable-eng/thanos/issues/1117

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"time"
 
 	"github.com/improbable-eng/thanos/pkg/objstore"
+	yaml "gopkg.in/yaml.v2"
```
please remove the blank line after this one

auhlig replied:
ran `goimports` again

povilasv replied:
goimports doesn't solve the blank line issue, please remove the blank line after this

auhlig replied:
i meant removed blank line and did the goimports again

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -65,6 +76,36 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+	// Support for cross domain scoping (user in different domain than project)
+	// If a userDomainName or userDomainID is given, a user is scoped to this domain
+	// and the tenant (aka project) is expected to be in the domain given by domainName or domainID.
+	if sc.UserDomainName != "" {
```
could we use switch here instead

```
switch {
case sc.UserDomaiName != "": 
...
case sc.UserDomainID != "":

}
```

auhlig replied:
right. done

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -65,6 +76,36 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+		}
+	}
+	if authOpts.Scope != nil {
+		if sc.TenantName != "" {
```
could we use switch here.

auhlig replied:
done

povilasv commented:
@auhlig you just commented, but didn't actually push the changes

auhlig commented:
@povilasv github was down this morning -.-

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+		level.Warn(logger).Log("msg", "usage of tenant_id is deprecated. use project_id instead")
+		authOpts.TenantID = sc.TenantID
+	case sc.ProjectName != "" || sc.ProjectID != "":
+		level.Info(logger).Log("msg", "using project_id or project_name")
```
I think we should remove this, it isn't a good message and IMO doesn't add any value

auhlig replied:
right. removed

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+		}
+	}
+	if authOpts.Scope != nil {
+		level.Info(logger).Log("msg", "scoping to domain, project")
```
same here let's not log this message.

auhlig replied:
removed

povilasv commented on pkg/objstore/swift/swift.go:
```diff
@@ -58,13 +69,57 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
 
+	// The term `tenant` is used in the deprecated OpenStack Identity v2.
+	// In Identity v3 the term `project` is used instead.
+	switch {
```
could you also add a couple of examples how or when to configure project id vs project name in the `docs/storage.md` file

auhlig replied:
updated docs. added link to official documentation with examples

auhlig commented:
LGTY @povilasv?

povilasv commented:
@auhlig Sort of yes. Did you look at the tests? Would be amazing if we could add some proper testing here, so that when we change code we don't break things.

bwplotka commented on pkg/objstore/swift/swift.go:
```diff
@@ -11,32 +11,43 @@ import (
+	RegionName        string `yaml:"region_name"`
+	ContainerName     string `yaml:"container_name"`
+
+	// Deprecated: Please use `project_id` instead.
```
I would maybe kill this and use `yaml.UmarshalStrict` instead? (: We already documented the change so let's fail fast on umarshalling! (:

auhlig replied:
sounds good. done

bwplotka commented on pkg/objstore/swift/swift.go:
```diff
@@ -58,13 +69,54 @@ func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+	// The term `tenant` is used in the deprecated OpenStack Identity v2.
+	// In Identity v3 the term `project` is used instead.
+	switch {
+	case sc.TenantName != "":
```
let's use `yaml.UnmarshalStrict` instead?

auhlig replied:
done

bwplotka commented on pkg/objstore/swift/swift.go:
```diff
@@ -180,13 +232,19 @@ func (c *Container) deleteContainer(name string) error {
+		UserDomainName:    os.Getenv("OS_USER_DOMAIN_NAME"),
+		ProjectDomainID:   os.Getenv("OS_PROJET_DOMAIN_ID"),
+		ProjectDomainName: os.Getenv("OS_PROJECT_DOMAIN_NAME"),
+		TenantID:          os.Getenv("OS_TENANT_ID"),
```
Kill this

auhlig replied:
killed

auhlig commented:
Made this a breaking change by dropping `tenant_id`, `tenant_name`. 
Extracted logic to `parseConfig`, `authOptsFromConfig` for easier unit testing.
Added unit tests.

LGTY @povilasv @bwplotka

auhlig commented:
Netlify doesn't seem to be in the mood to run the tests. CircleCI passed however.

auhlig commented:
squashed into single commit and force pushed.
giving netlify another chance

auhlig commented:
Updated the changelog after some merge conflicts. Can this be merged @povilasv @bwplotka?

GiedriusS commented:
@auhlig so essentially with this PR we are getting rid of the support of the V2 API? I'm a bit anxious about this since, for example, some companies might still be running old versions of that API (*cough* not gonna point any fingers). Would it be hard to add some kind of backward compatibility?

auhlig commented:
@GiedriusS This PR does **not** drop Identity v2 support. It's still possible to authenticate using `project_id` or `project_name` instead of the deprecated terms `tenant_id`, `tenant_name`.
We only fix the terminology ( tenant became project with Identity v3) and add support for for authentication if a  user is in a different domain than the project.
In summary, the only breaking change is the terminology.

povilasv commented:
Ok so this does include a breaking change
`tenantID`, `tenantName` -> `projectID`, `projectName`.

I think in order not to break users, just fill tenantID to projectID and tenantName to projectName.

We try to avoid breaking changes when possible and in this case it's definitely doable

bwplotka commented:
@povilasv I think breaking here is better that having inconsistent configuration -> users will be immdiately notified by verbose error message.

bwplotka commented:
Waiting for @povilasv final LGTM to merge (:

auhlig commented:
Thank you guys 🎉
